syntax = "proto2";

package ydec.mahjong;

import "mahjong-common.proto";

message Rule {
  repeated Yaku yaku = 1;
}

message Yaku {
  optional string name = 1;
  optional string description = 2;
  optional int32 menzen_fan = 3;
  optional int32 kuisagari_fan = 4;
  optional int32 force_fu = 5;
  optional YakuCondition yaku_condition = 6;

  // Upper version of this yaku.
  // This yaku will not be applied if there is any upper version yaku for the given hand.
  repeated string upper_version_yaku_name = 7;
}

message TileCondition {
  enum VariableTileType {
    MASK_VARIABLE_TYPE = 0xf0;

    // VARIABLE_TILE_A to VARIABLE_TILE_G are all distinct each other.
    VARIABLE_TILE = 0x10;
      VARIABLE_TILE_A = 0x11;
      VARIABLE_TILE_B = 0x12;
      VARIABLE_TILE_C = 0x13;
      VARIABLE_TILE_D = 0x14;
      VARIABLE_TILE_E = 0x15;
      VARIABLE_TILE_F = 0x16;
      VARIABLE_TILE_G = 0x17;

    // VARIABLE_TILE2_A to VARIABLE_TILE2_C are all distinct each other.
    VARIABLE_TILE2 = 0x20;
      VARIABLE_TILE2_A = 0x21;
      VARIABLE_TILE2_B = 0x22;
      VARIABLE_TILE2_C = 0x23;

    // VARIABLE_NUMBER_A to VARIABLE_NUMBER_C are all distinct each other.
    VARIABLE_NUMBER = 0x30;
      VARIABLE_NUMBER_A = 0x31;
      VARIABLE_NUMBER_B = 0x32;
      VARIABLE_NUMBER_C = 0x33;

    VARIABLE_COLOR = 0x40;
      // Doesn't include Jihai. (Jihai-tile doesn't have any color.)
      VARIABLE_COLOR_A = 0x41;
      VARIABLE_COLOR_A_OR_JIHAI = 0x42;
  }
  optional TileType required_tile = 1;
  optional VariableTileType required_variable_tile = 2;
  repeated TileState required_state = 3;
  repeated TileState disallowed_state = 4;
}

message ElementCondition {
  repeated HandElementType allowed_element_type = 1;

  // You may *not* use two or more variable_tiles.
  repeated TileCondition allowed_tile_condition = 2;

  repeated TileCondition required_tile_condition = 3;

  // At least one of the given conditions must be satisfied.
  // You may *not* use variable_tile.
  repeated TileCondition either_tile_condition = 4;
}

message AgariCondition {
  optional AgariType required_type = 1;
  repeated AgariState required_state = 2;
  repeated AgariFormat allowed_format = 3;
}

message YakuCondition {
  // You may *not* use two or more variable_tiles.
  repeated TileCondition allowed_tile_condition = 1;

  // You may *not* use variable_tile.
  repeated TileCondition disallowed_tile_condition = 2;

  repeated TileCondition required_tile_condition = 3;
  repeated ElementCondition required_element_condition = 4;
  optional MachiType required_machi_type = 5;
  optional PlayerType required_player_type = 6;
  optional AgariCondition required_agari_condition = 7;
  optional RichiType required_richi_type = 8;
  optional TileType required_field_wind = 9;
  optional TileType required_player_wind = 10;
}
